File: pybot/state.py
# -*- coding: utf-8 -*-
"""JSON-file state store for the bot.

Holds per-chat profiles (last callsign / grid square / language), in-flight
wizard sessions and the cached satellite catalog. All writes are atomic
(tmp file + os.replace) and guarded by a module-level lock.
"""

import json
import os
import threading
import time

_LOCK = threading.Lock()


def _load(path):
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
        if isinstance(data, dict):
            return data
    except (OSError, ValueError):
        pass
    return {}


def _save(path, data):
    tmp = path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    os.replace(tmp, path)


def _ensure_sections(data):
    data.setdefault("profiles", {})
    data.setdefault("wizards", {})
    data.setdefault("catalog", {})
    return data


class Store:
    def __init__(self, path):
        self.path = path

    def _data(self):
        data = _load(self.path)
        return _ensure_sections(data)

    def _write(self, fn):
        with _LOCK:
            data = self._data()
            fn(data)
            _save(self.path, data)

    # ---- profiles ---------------------------------------------------

    def get_profile(self, chat_id):
        with _LOCK:
            return self._data()["profiles"].get(str(chat_id), {})

    def save_profile(self, chat_id, profile):
        def upd(data):
            data["profiles"][str(chat_id)] = profile

        self._write(upd)

    # ---- wizards ----------------------------------------------------

    def get_wizard(self, chat_id):
        with _LOCK:
            return self._data()["wizards"].get(str(chat_id))

    def save_wizard(self, wizard):
        def upd(data):
            data["wizards"][str(wizard["chat_id"])] = wizard

        self._write(upd)

    def clear_wizard(self, chat_id):
        def upd(data):
            data["wizards"].pop(str(chat_id), None)

        self._write(upd)

    # ---- catalog cache ----------------------------------------------

    def get_catalog(self):
        with _LOCK:
            entry = self._data()["catalog"]
        if not entry:
            return None
        if entry.get("ts", 0) + entry.get("ttl", 0) < time.time():
            return None
        return entry.get("data")

    def set_catalog(self, sats, ttl):
        def upd(data):
            data["catalog"] = {"ts": int(time.time()), "ttl": ttl, "data": sats}

        self._write(upd)

File: tgbot/src/state.ts
import { KV_KEYS, WIZARD_TTL } from "./constants";
import type { Env, Profile, WizardState } from "./types";

export async function getWizard(env: Env, chatId: number): Promise<WizardState | null> {
  const raw = await env.SESSIONS.get(`${KV_KEYS.wizard}${chatId}`);
  return raw ? (JSON.parse(raw) as WizardState) : null;
}

export async function saveWizard(env: Env, state: WizardState): Promise<void> {
  await env.SESSIONS.put(`${KV_KEYS.wizard}${state.chatId}`, JSON.stringify(state), {
    expirationTtl: WIZARD_TTL,
  });
}

export async function clearWizard(env: Env, chatId: number): Promise<void> {
  await env.SESSIONS.delete(`${KV_KEYS.wizard}${chatId}`);
}

export async function getProfile(env: Env, chatId: number): Promise<Profile> {
  const raw = await env.SESSIONS.get(`${KV_KEYS.profile}${chatId}`);
  if (raw) {
    try {
      return JSON.parse(raw) as Profile;
    } catch {
      /* fall through */
    }
  }
  return { lang: "en" };
}

export async function saveProfile(env: Env, chatId: number, profile: Profile): Promise<void> {
  await env.SESSIONS.put(`${KV_KEYS.profile}${chatId}`, JSON.stringify(profile));
}